pkg/agent/util/net_windows.go

```diff
@@ -613,92 +643,104 @@ func HostInterfaceExists(ifaceName string) bool {
+			}
 		}
 	}
+	return nil
+}
 
-	if err := RemoveNetRoute(route); err != nil {
-		return err
+func ReplaceNetRoute(route *Route) error {
+	if route == nil || route.DestinationSubnet == nil {
+		return nil
 	}
-	if err := NewNetRoute(route); err != nil {
-		return err
+	family := addressFamilyByIP(route.DestinationSubnet.IP)
+	rows, err := listIPForwardRows(antreaNetIO, family)
+	if err != nil {
+		return fmt.Errorf("unable to list Windows IPForward rows: %v", err)
+	}
+	for i := range rows {
+		row := rows[i]
+		if row.DestinationPrefix.EqualsTo(route.DestinationSubnet) && row.Index == uint32(route.LinkIndex) {
+			if row.NextHop.IP().Equal(route.GatewayAddress) {
+				return nil
+			} else {
+				if err := antreaNetIO.DeleteIPForwardEntry(&row); err != nil {
+					return fmt.Errorf("failed to delete existing route with nextHop %s: %v", route.GatewayAddress, err)
+				}
+			}
+		}
 	}
-	return nil
+	return NewNetRoute(route)
 }
 
 func GetNetRoutes(linkIndex int, dstSubnet *net.IPNet) ([]Route, error) {
-	cmd := fmt.Sprintf("Get-NetRoute -InterfaceIndex %d -DestinationPrefix %s -ErrorAction Ignore | Format-Table -HideTableHeaders",
-		linkIndex, dstSubnet.String())
-	return getNetRoutes(cmd)
+	if dstSubnet == nil {
+		return nil, fmt.Errorf("unable to get net routes for %d", linkIndex)
+	}
+	family := addressFamilyByIP(dstSubnet.IP)
+	return listRoutes(family, func(row antreasyscall.MibIPForwardRow) bool {
+		return row.DestinationPrefix.EqualsTo(dstSubnet) && row.Index == uint32(linkIndex)
+	})
 }
 
 func GetNetRoutesAll() ([]Route, error) {
-	cmd := "Get-NetRoute -ErrorAction Ignore | Format-Table -HideTableHeaders"
-	return getNetRoutes(cmd)
+	return listRoutes(antreasyscall.AF_UNSPEC, nil)
 }
 
-func getNetRoutes(cmd string) ([]Route, error) {
-	routesStr, _ := runCommand(cmd)
-	parsed := parseGetNetCmdResult(routesStr, 6)
-	var routes []Route
-	for _, items := range parsed {
-		idx, err := strconv.Atoi(items[0])
-		if err != nil {
-			return nil, fmt.Errorf("failed to parse the LinkIndex '%s': %v", items[0], err)
-		}
-		_, dstSubnet, err := net.ParseCIDR(items[1])
-		if err != nil {
-			return nil, fmt.Errorf("failed to parse the DestinationSubnet '%s': %v", items[1], err)
-		}
-		gw := net.ParseIP(items[2])
-		metric, err := strconv.Atoi(items[3])
-		if err != nil {
-			return nil, fmt.Errorf("failed to parse the RouteMetric '%s': %v", items[3], err)
-		}
-		route := Route{
-			LinkIndex:         idx,
-			DestinationSubnet: dstSubnet,
-			GatewayAddress:    gw,
-			RouteMetric:       metric,
+type routeFilter func(row antreasyscall.MibIPForwardRow) bool
+
+func listRoutes(family uint16, filter routeFilter) ([]Route, error) {
+	rows, err := listIPForwardRows(antreaNetIO, family)
+	if err != nil {
+		return nil, fmt.Errorf("unable to list Windows IPForward rows: %v", err)
+	}
+	rts := make([]Route, 0, len(rows))
+	for _, r := range rows {
+		if filter != nil || filter(r) {
```

Comment:
I think it should be `filter == nil || filter(r)`, otherwise all routes would be returned when any filter is provided.
We need unit test to validate the function.

Reply:
Added UT for `GetNetRoutesAll` where filter is nil.